dysnix/bsc/templates/scripts/_init_from_gcs.tpl

```diff
@@ -97,45 +98,77 @@ TIMESTAMP_0="$(${S5CMD} cat s3://${TIMESTAMP_URL})"
 
 # we're ready to perform actual data sync
 
-# we're done when both are true
+# we're done when all are true
 # 1) start and stop timestamps did not changed during data sync - no process started or finished updating the cloud
-# 2) 0 objects copied
+# 2) start timestamp is before stop timestamp - no process is in progress updating the cloud
+# 3) 0 objects copied
 SYNC=2
 CLEANUP=1
 while [ "${SYNC}" -gt 0 ] ; do
 
     # Cleanup
     if [ ${CLEANUP} -eq 1 ];then
-      echo "$(date -Iseconds) Cleaning up local dir ..."
-      mkdir -p ${DATA_DIR}/geth
-      mv ${DATA_DIR}/geth ${DATA_DIR}/geth.old && rm -rf ${DATA_DIR}/geth.old &
+      echo "$(date -Iseconds) Cleaning up local dir ${GETH_DIR} ..."
+      mkdir -p "${GETH_DIR}"
+      mv "${GETH_DIR}" "${GETH_DIR}.old" && rm -rf "${GETH_DIR}.old" &
       CLEANUP=0
     fi
 
-    # sync from cloud to local disk, without removing existing [missing in the cloud] files
+    # sync from cloud to local disk, with removing existing [missing in the cloud] files
     # run multiple syncs in background
 
-    # we don't wanna sync ancient data here
-    time ${S5CMD} sync ${EXCLUDE_ANCIENT} s3://${STATE_SRC}/* ${CHAINDATA_DIR}/ > cplist_state.txt &
+    time ${S5CMD} sync --delete s3://${STATE_SRC}/* ${STATE_TMP_DIR}/ > cplist_state.txt &
     STATE_CP_PID=$!
-    time nice ${S5CMD} sync --delete --part-size 200 --concurrency 2 ${EXCLUDE_STATE} s3://${ANCIENT_SRC}/* ${CHAINDATA_DIR}/ancient/ > cplist_ancient.txt &
+    time nice ${S5CMD} sync --delete --part-size 200 --concurrency 2 s3://${ANCIENT_SRC}/* ${ANCIENT_TMP_DIR}/ > cplist_ancient.txt &
     ANCIENT_CP_PID=$!
 
     # wait for all syncs to complete
-    # TODO any errors handling here?
-    wait ${STATE_CP_PID} ${ANCIENT_CP_PID}
-
+    # shell tracks all sub-processes and stores exit codes internally
+    # it's not required to stay in wait state for all background processes at the same time
+    # we'll handle these processes sequentially
+    wait ${STATE_CP_PID}
+    STATE_CP_EXIT_CODE=$?
+    wait ${ANCIENT_CP_PID}
+    ANCIENT_CP_EXIT_CODE=$?
+
+    # let's handle out of disk space specially, thus we don't re-try, just stuck here if disk usage is high
+    VOLUME_USAGE_PERCENT=$(df "${DATA_DIR}" | tail -n 1 | awk '{print $5}'|tr -d %)
+    if [ "${VOLUME_USAGE_PERCENT}" -gt "${MAX_USED_SPACE_PERCENT}" ];then
+      set +x
+      # stop monitoring
+      if [ ${MON_PID} -ne 0 ];then kill ${MON_PID};MON_PID=0; fi
+      echo "We're out of disk space. Stuck here, nothing we can do. Check the source snapshot size"
+      echo "Source snapshot size ${REMOTE_STATS}"
+      echo "Disk usage is ${VOLUME_USAGE_PERCENT}%"
+      df -P -BG "${DATA_DIR}"
+      # we need to sleep <short-delay> inside loop to handle pod termination w/o delays
+      # infinite sleep loop
+      while true; do sleep 10;done
```

Comment:
why do we want to stuck here instead of exiting with error?

Reply:
Well, the implementation cleanі up everything on init container re-exec, it's intended to be used on fresh nodes with empty disks.  Thus, using "exit on disk full" we are going to stay in the loop to downloading the data from S3 all the time, and it's billable. Taking into amount that it takes like 30-40 minutes in my case - it will not be detected as crash loopback. It seems the correct option should be to flag out-of-space with an empty file ( I hope we're out of space, not inodes) and then perform exit on error. Plus to check this file on some early stage.  I would say "stuck forever" is just first option I was able to write. Will re-work this with a "lock" file.

Reply:
Thats reasonable, thanks for explaining

Reply:
@VladStarr I've re-worked out-of-space handling using an approach described above.